paddle/fluid/operators/filter_instag_op.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/operators/filter_instag_op.h"
+
+#include <memory>
+#include "paddle/fluid/framework/no_need_buffer_vars_inference.h"
+#include "paddle/fluid/framework/var_type_inference.h"
+
+namespace paddle {
+namespace operators {
+class FilterInstagOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X1"), "Input(X1) should be not null.");
+    PADDLE_ENFORCE(ctx->HasInput("X2"), "Input(X2) should be not null.");
+    PADDLE_ENFORCE(ctx->HasInput("X3"), "Input(X3) should be not null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"), "Output(Out) should be not null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Map"), "Output(Map) should be not null.");
+
+    if (!ctx->IsRuntime()) {
+      auto x1_dims = ctx->GetInputDim("X1");  // batch_size * vec
+
+      ctx->SetOutputDim("Out", framework::make_ddim({-1, x1_dims[1]}));
+      ctx->SetOutputDim("Map", framework::make_ddim({-1, 2}));
+
+    } else {
+      auto x1_dims = ctx->GetInputDim("X1");
+      ctx->SetOutputDim("Out", framework::make_ddim({x1_dims[0], x1_dims[1]}));
+      // 3 number is start line of Out, start line of X1, and line counts
+      ctx->SetOutputDim("Map", framework::make_ddim({x1_dims[0], 3}));
+    }
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext& ctx) const override {
+    auto data_type = framework::GetDataTypeOfVar(ctx.InputVar("X1"));
+    return framework::OpKernelType(data_type, ctx.device_context());
+  }
+};
+
+class FilterInstagOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("X1", "(LoDTensor) global fc output");
```

Comment:
description of X1, X2, X3 is not readable. global fc is confused.

Reply:
Thank you.
x1 is ins, x2 is ins_tag, x3 is filter_tag. they have been renamed.